pkg/query/querier.go

```diff
@@ -208,7 +208,19 @@ func (q *querier) Select(_ bool, hints *storage.SelectHints, ms ...*labels.Match
 	// The querier has a context but it gets canceled, as soon as query evaluation is completed, by the engine.
 	// We want to prevent this from happening for the async storea API calls we make while preserving tracing context.
 	ctx := tracing.CopyTraceContext(context.Background(), q.ctx)
-	ctx, cancel := context.WithTimeout(ctx, q.selectTimeout)
+
+	perRequestCancelTime, deadlineSpecified := q.ctx.Deadline()
+
+	var perQueryTimeOut time.Duration
+
+	// If the per-request timeout is specified, the query would timeout per the specified per-request timeout value.
+	// If the per-request timeout is not specified, the query would timeout per the timeout value specified at global query component.
+	if deadlineSpecified {
```

Comment:
As I commented before: Not sure if we need this. We depend on q.ctx then we create another context with another timeout from that context and we use it. This mean that the whatever deadline was shorter, it will be triggered

Reply:
What I am saying is that we can do just

```
ctx, cancel := context.WithTimeout(ctx, perQueryTimeOut)
ctx, cancel := context.WithTimeout(ctx, perRequestCancelTime)
```
Or maybe for minor "efficiency" (one goroutine less)

```
timeout := time.Until(perRequestCancelTime)
if timeout > q.selectTimeout {
    timeout = q.selectTimeout
}
```

Reply:
This check (if deadlineSpecified) helps two distinct possibilities: (1) User might not specify any per-request timeout value as its optional (2) User might wanna wait longer time than default query component level timeout

Reply:
Cool, I think I missed (2) totally. The only concern I have is that it's common missed idea and we might have enforcement of some default timeouts in other places as well. Let's add ticket for this / docs.